ssl/ssl_lib.c

```diff
@@ -3774,90 +3774,82 @@ SSL *SSL_dup(SSL *s)
-        }
+    /* Fix up some values */
+    ret->references = 1;
+    ret->rlayer.s = ret;
 
-        if (!SSL_set_session_id_context(ret, s->sid_ctx,
-                                        (int)s->sid_ctx_length))
-            goto err;
-    }
+    /*
+     * Ensure any malloc'd fields are NULL so we can safely call SSL_free().
+     * Fields that get malloc'd after the handshake has started can be ignored
+     * because we shouldn't get that far in that case.
+     */
+    ret->rbio = ret->wbio = NULL;
+    ret->s3 = NULL;
+    ret->d1 = NULL;
+    ret->param = NULL;
+    /* we use ssl_dane_dup() for this structure */
+    memset(&ret->dane, 0, sizeof(SSL_DANE));
+    ret->cipher_list = NULL;
+    ret->cipher_list_by_id = NULL;
+    ret->tls13_ciphersuites = NULL;
+    ret->cert = NULL;
+    ret->session = NULL;
+    ret->psksession = NULL;
+    /* We use CRYPTO_dup_ex_data() for this structure */
+    memset(&ret->ex_data, 0, sizeof(CRYPTO_EX_DATA));
+    ret->ctx = NULL;
+    ret->ca_names = ret->client_ca_names = NULL;
+# ifndef OPENSSL_NO_EC
+    ret->ext.ecpointformats = NULL;
+# endif                         /* OPENSSL_NO_EC */
+    ret->ext.supportedgroups = NULL;
+    ret->ext.alpn = NULL;
+    ret->session_ctx = NULL;
+    ret->srtp_profiles = NULL;
+    /* We use ssl_srp_ctx_init_int() for this structure */
+    memset(&ret->srp_ctx, 0, sizeof(SRP_CTX));
+    s->rlayer.d = NULL;
+    ret->lock = NULL;
+
+    if (!SSL_CTX_up_ref(s->ctx))
+        goto err;
+    ret->ctx = s->ctx;
 
-    if (!ssl_dane_dup(ret, s))
+    /* setup rbio, and wbio */
+    if (s->rbio != NULL && !BIO_up_ref(s->rbio))
         goto err;
-    ret->version = s->version;
-    ret->options = s->options;
-    ret->mode = s->mode;
-    SSL_set_max_cert_list(ret, SSL_get_max_cert_list(s));
-    SSL_set_read_ahead(ret, SSL_get_read_ahead(s));
-    ret->msg_callback = s->msg_callback;
-    ret->msg_callback_arg = s->msg_callback_arg;
-    SSL_set_verify(ret, SSL_get_verify_mode(s), SSL_get_verify_callback(s));
-    SSL_set_verify_depth(ret, SSL_get_verify_depth(s));
-    ret->generate_session_id = s->generate_session_id;
-
-    SSL_set_info_callback(ret, SSL_get_info_callback(s));
-
-    /* copy app data, a little dangerous perhaps */
-    if (!CRYPTO_dup_ex_data(CRYPTO_EX_INDEX_SSL, &ret->ex_data, &s->ex_data))
+    ret->rbio = s->rbio;
+    if (s->wbio != NULL && !BIO_up_ref(s->wbio))
         goto err;
+    ret->wbio = s->wbio;
 
-    /* setup rbio, and wbio */
-    if (s->rbio != NULL) {
-        if (!BIO_dup_state(s->rbio, (char *)&ret->rbio))
-            goto err;
-    }
-    if (s->wbio != NULL) {
-        if (s->wbio != s->rbio) {
-            if (!BIO_dup_state(s->wbio, (char *)&ret->wbio))
-                goto err;
-        } else {
-            BIO_up_ref(ret->rbio);
-            ret->wbio = ret->rbio;
-        }
-    }
+    ret->s3 = OPENSSL_malloc(sizeof(*ret->s3));
+    if (ret->s3 == NULL)
+        goto err;
+    *ret->s3 = *s->s3;
```

Comment:
Will it be problem to assign `*s->s3` to `*ret->s3` directly since there are some pointers in `*s->s3 (struct ssl3_state_st)`, if so there will be multiple variables pointing to the same memory. 
When processing `s->rlayer.d` below, we call `DTLS_RECORD_LAYER_new` to create new DTLS_RECORD_LAYER struct, but not simply assign `*s->rlayer.d` to `*rec->rlayer.d`.
Similar question to `*ret->d1 = *s->d1` below.
What is the scenario of using `SSL_dup`, I'm still confused in fact.

Reply:
> Will it be problem to assign `*s->s3` to `*ret->s3` directly since there are some pointers in `*s->s3 (struct ssl3_state_st)`, if so there will be multiple variables pointing to the same memory.

We can safely ignore all of those pointers because of this code at the beginning of `SSL_dup`:
````
    /* If we're not quiescent, just up_ref! */
    if (!SSL_in_init(s) || !SSL_in_before(s)) {
        CRYPTO_UP_REF(&s->references, &i, s->lock);
        return s;
    }
````
i.e. `SSL_dup` is a bit weird because if the SSL is in use then it just ups the ref count and doesn't dup at all. So the only time we *really* duplicate the SSL object is if we're doing so *before* the handshake has started. The pointers in `s->s3` only get populated during the handshake so will be NULL if we get to this point in the code.

> When processing `s->rlayer.d` below, we call `DTLS_RECORD_LAYER_new` to create new DTLS_RECORD_LAYER struct, but not simply assign `*s->rlayer.d` to `*rec->rlayer.d` (or else simply assign `s->rlayer.d` to `rec->rlayer.d`).

Again this is because we are duplicating before the handshake has started. The DTLS1_RECORD_LAYER struct does not have anything interesting in it before the handshake has started so we just create a new one. We could equally have malloc'd the space and then copied the as yet unused data across. But since we have a `DTLS_RECORD_LAYER_new` function we might as well use it.

> Similar question to `*ret->d1 = *s->d1` below.

As above we can safely do this assignment because we know we are doing this before the handshake has started.